### PR TITLE
reynolds number

### PR DESCRIPTION
**Pull request by aokiyoshi**
first commit

### CODE REVIEW

File: examples/hydro/reynolds_num_example.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from symplyphysics.laws.hydro import reynolds_number
+from symplyphysics import units, convert_to, dimensionless
+
+rho = 1000 * units.kilogram / units.meter**3
```
Comment (blackyblack) [inline examples/hydro/reynolds_num_example.py:4]: Quantities are not defined this way. See other examples for reference.
Also example is not required for the test task.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+from sympy import (Eq, solve)
+from symplyphysics import (
+    units, Quantity, Symbol, print_expression, validate_input, validate_output, dimensionless)
+
+
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:5]: Law should have a description. See other laws for reference.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+diameter = Symbol("d", units.length)  # hydraulic diameter of the pipe
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:6]: Do not shortcut. Use full names for symbols.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+# density of the fluid kg/m^3
+density = Symbol("rho", units.mass / units.volume)
+velocity = Symbol("v", units.length / units.time)  # velocity of the fluid m/s
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:9]: Density and velocity should be available in units. Check them out.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@validate_input(
+    diameter_=diameter,
+    density_=density,
+    velocity_=velocity,
+    dynamic_viscosity_=dynamic_viscosity
+)
+@validate_output(reynolds_number)
+def calculate_reynolds_number(diameter_, density_, velocity_, dynamic_viscosity_):
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:28]: Use types for parameters and returned value of a function.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    result_reynolds_number_expr = solve(law, reynolds_number, dict=True)[
+        0][reynolds_number]
+    result_expr = result_reynolds_number_expr.subs({
+        diameter: diameter_,
+        density: density_,
+        velocity: velocity_,
+        dynamic_viscosity: dynamic_viscosity_
+    })
+    return Quantity(result_expr, dimension=dimensionless)
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:37]: Do not override dimension. It should be deducted automatically.

File: test/hydro/reynolds_number_test.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    units
+)
+from symplyphysics.laws.hydro import reynolds_number
+
+@fixture(name="test_args")
+def test_args_fixture():
+    rho = Quantity(1000 * units.kilogram / units.meter**3)
+    d = Quantity(0.1 * units.meter)
+    v = Quantity(1 * units.meter / units.second)
+    mu = Quantity(0.000894 * units.pressure * units.time)
```
Comment (blackyblack) [inline test/hydro/reynolds_number_test.py:17]: Quantities use unit of measurement, not dimension in parameters.

File: test/hydro/reynolds_number_test.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    Args = namedtuple("Args", ["rho", "d", "v", "mu"])
+    return Args(rho=rho, d=d, v=v, mu=mu)
+
+def test_reynolds_number(test_args):
+    result = reynolds_number.calculate_reynolds_number(test_args.d, test_args.rho, test_args.v, test_args.mu)
+    assert SI.get_dimension_system().equivalent_dims(result.dimension, dimensionless)
+    result_reynolds_number = convert_to(result, dimensionless).evalf(3)
+    assert result_reynolds_number == approx(100000, 0.001)
```
Comment (blackyblack) [inline test/hydro/reynolds_number_test.py:25]: This is not a vaild result. Use calculator to check.
So your test is synthetic - it is using fake result to verify, therefore it does not check anything.

**Review by blackyblack — CHANGES_REQUESTED** [review]

File: test/hydro/reynolds_number_test.py (hunk 8 above)
Comment (aokiyoshi) [reply test/hydro/reynolds_number_test.py]: fixed

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
+from sympy import (Eq, solve)
+from symplyphysics import (
+    units, Quantity, Symbol, print_expression, validate_input, validate_output, dimensionless)
+
+
+# Description
+# This file contains the implementation of the Reynolds number calculation in fluid dynamics.
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:7]: Remove this paragraph.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+## The Reynolds number quantifies the relative importance of these two types of forces for given flow conditions,
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:8]: I do not see, which two types of forces are referenced here. Please rephrase.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+## and is a guide to when turbulent flow will occur in a particular situation
+# The Reynolds number is a dimensionless quantity that characterizes the flow of a fluid in a pipe.
+# It is calculated using the hydraulic diameter, fluid density, fluid velocity, and dynamic viscosity.
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:11]: Remove this paragraph. It is a duplicate of law definition below.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+# Law: Re = rho * d * v / mu, where
+# rho is fluid density
+# d is hydraulic diameter of the pipe
+# v is velocity of the fluid m/s.
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:15]: Remove `m/s`. It is preferable not to use concrete units in description.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+
+diameter = Symbol("diameter", units.length)  # hydraulic diameter of the pipe
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:18]: Remove comments. They are duplicate of the description.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+density = Symbol(
+    "density", (units.mass / units.volume)  # density of the fluid kg/m^3
+)
+velocity = Symbol("volume", units.velocity)  # velocity of the fluid m/s
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:22]: Wrong name.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+# It is calculated using the hydraulic diameter, fluid density, fluid velocity, and dynamic viscosity.
+# Law: Re = rho * d * v / mu, where
+# rho is fluid density
+# d is hydraulic diameter of the pipe
+# v is velocity of the fluid m/s.
+
+
+diameter = Symbol("diameter", units.length)  # hydraulic diameter of the pipe
+density = Symbol(
+    "density", (units.mass / units.volume)  # density of the fluid kg/m^3
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:20]: Did you check if `units.density` is available?

File: test/hydro/reynolds_number_test.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+# Description
+## There is pipe with diameter 0.1m, density 1000kg/m3, velocity 1m/s
+## and dynamic viscosity 0.000894Pa*s. The reynolds number should be 111856.823
+
+@fixture(name="test_args")
+def test_args_fixture():
+    rho = Quantity(1000 * units.kilogram / units.meter**3)
+    d = Quantity(0.1 * units.meter)
+    v = Quantity(1 * units.velocity)
```
Comment (blackyblack) [inline test/hydro/reynolds_number_test.py:20]: Velocity is dimension. You should use units of measurement here.

File: examples/hydro/reynolds_num_example.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
+from symplyphysics.laws.hydro import reynolds_number
```
Comment (blackyblack) [inline examples/hydro/reynolds_num_example.py:1]: Remove example from this PR. It is a separate task.

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+@validate_input(
+    diameter_=diameter,
+    density_=density,
+    velocity_=velocity,
+    dynamic_viscosity_=dynamic_viscosity
+)
+@validate_output(reynolds_number)
+def calculate_reynolds_number(diameter_: Quantity, density_: Quantity,
+    velocity_: Quantity, dynamic_viscosity_: Quantity) -> float:
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:45]: You write the returned type is float but return Quantity. Check out other laws, how dimensionless result is treated.

File: symplyphysics/laws/hydro/reynolds_number.py (hunk 15 above)
Comment (aokiyoshi) [reply symplyphysics/laws/hydro/reynolds_number.py:20]: I did. It's not

File: symplyphysics/laws/hydro/reynolds_number.py (hunk 18 above)
Comment (aokiyoshi) [reply symplyphysics/laws/hydro/reynolds_number.py]: fixed

File: test/hydro/reynolds_number_test.py [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+def test_args_fixture():
+    rho = Quantity(1000 * units.kilogram / units.meter**3)
+    d = Quantity(0.1 * units.meter)
+    v = Quantity(1 * units.meter / units.second)
+    mu = Quantity(0.000894 * units.pascal * units.second)
+    Args = namedtuple("Args", ["d", "rho", "v", "mu"])
+    return Args(d=d, rho=rho, v=v, mu=mu)
+
+
+def test_reynolds_number(test_args):
```
Comment (blackyblack) [inline test/hydro/reynolds_number_test.py:21]: Where are bad input parameters tests? See other tests for reference.

**Review by blackyblack — APPROVED** [review]

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+                              velocity_: Quantity, dynamic_viscosity_: Quantity) -> float:
+    result_expr = solve(law, reynolds_number, dict=True)[0][reynolds_number]
+    result_applied = result_expr.subs({
+        diameter: diameter_,
+        density: density_,
+        velocity: velocity_,
+        dynamic_viscosity: dynamic_viscosity_
+    })
+    result = Quantity(result_applied)
+    return float(convert_to(result, dimensionless))
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:49]: Type checker error:
https://github.com/blackyblack/symplyphysics/actions/runs/7339956644/job/19985194465?pr=131

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+    result_expr = solve(law, reynolds_number, dict=True)[0][reynolds_number]
+    result_applied = result_expr.subs({
+        diameter: diameter_,
+        density: density_,
+        velocity: velocity_,
+        dynamic_viscosity: dynamic_viscosity_
+    })
+    result = Quantity(result_applied)
+    
+    return float(convert_to(result, Quantity(dimensionless)))
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:50]: While it works, it is not a correct way of converting from dimensionless quantities. Here you take a Dimension, convert it to Quantity and later to sympy Quantity.
See how it is done properly: https://github.com/blackyblack/symplyphysics/blob/17d00b153fc87168dba61bdf9eb94ca53dd2d14f/symplyphysics/definitions/refractive_index_is_wave_speeds_ratio.py#L49

File: symplyphysics/laws/hydro/reynolds_number.py [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    result_applied = result_expr.subs({
+        diameter: diameter_,
+        density: density_,
+        velocity: velocity_,
+        dynamic_viscosity: dynamic_viscosity_
+    })
+    result = Quantity(result_applied)
+    
+    return float(convert_to(result, S.One).evalf())
+
```
Comment (blackyblack) [inline symplyphysics/laws/hydro/reynolds_number.py:51]: Run pylint --fail-under=9.8 ./test/**/*.py ./symplyphysics
************* Module reynolds_number_test
test/hydro/reynolds_number_test.py:35:0: C0303: Trailing whitespace (trailing-whitespace)
test/hydro/reynolds_number_test.py:45:0: C0303: Trailing whitespace (trailing-whitespace)
test/hydro/reynolds_number_test.py:65:0: C0305: Trailing newlines (trailing-newlines)
************* Module symplyphysics.laws.hydro.reynolds_number
symplyphysics/laws/hydro/reynolds_number.py:49:0: C0303: Trailing whitespace (trailing-whitespace)
symplyphysics/laws/hydro/reynolds_number.py:51:0: C0305: Trailing newlines (trailing-newlines)

-----------------------------------
Your code has been rated at 9.99/10

**Review by blackyblack — APPROVED** [review]

**Pull request merged**